packages/clay-shared/src/useDebounce.ts

```diff
@@ -11,15 +11,24 @@ import {useEffect, useState} from 'react';
 const useDebounce = (value: any, delay: number) => {
 	const [debouncedValue, setDebouncedValue] = useState(value);
 
-	useEffect(() => {
-		const handler = setTimeout(() => {
-			setDebouncedValue(value);
-		}, delay);
+	useEffect(
+		() => {
+			const handler = setTimeout(() => {
+				setDebouncedValue(value);
+			}, delay);
 
-		return () => {
-			clearTimeout(handler);
-		};
-	}, [value]);
+			return () => {
+				clearTimeout(handler);
+			};
+		},
+		// This is required when the `object` has lost the
+		// reference plus the values are the same, `useEffect`
+		// uses `Object.is` or equivalent underneath the cloths.
+		// For some reason the reference is being lost.
+		typeof value === 'object'
+			? [...Object.keys(value), ...Object.values(value)]
```

Comment:
I'm not sure I get what is happening here. The array you are constructing is also an object so I would expect `Object.is` to treat it just like `value`... It would be good to really understand what's happening.

Reply:
@wincent I spent a few hours debugging the core of React to understand what was happening on Friday, it's pretty much what I'm guessing, possibly the value is missing the reference because within React it stores within the Fibers the effects with the inputs of the previous and next effects, in this case the array with the `value` that is an object... to call the `useEffect` it check if there are inputs in [updateEffectImpl](https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberHooks.js#L870-L889) and then below it decides whether to throw the effect or not checking if the inputs are equal or not, in the end it compares each array item within the implementation [areHookInputsEqual](https://github.com/facebook/react/blob/master/packages/react-dom/src/server/ReactPartialRendererHooks.js#L112-L118) which [in the end uses a polyfill of Object.is](https://github.com/facebook/react/blob/master/packages/shared/objectIs.js#L14-L18) comparing the value of the old and new object. What happened was that the values ​​were the same and did not change the order but in the end it was not the same reference when you compare `a === b`.

I'm not 100% sure if the reference loss is going on inside React for some reason or it's on our side.

Reply:
Got it. That makes sense. So given an array it is going to compare the values rather than comparing the containing array itself. And because you are using `Object.keys` (ie. immutable strings) and `Object.values` (ie. references), so they end up always being equivalent, unless one of the `key` or `value` items change.

I'm not 100% sure what you mean by "missing the reference" or "lost the reference", but you just mean the object identity has changed, right?

Reply:
Yes, this may be more of a hack using `Object.values` and `Object.keys` but it's really what I wanted here, just dribbling the identity loss but verifying that the keys and values ​​still remain the same.

I tried to check if it was a problem in React, but it's a bit tricky to understand all the ways, I would spend a few days 😅... but I'm also not sure if `Object.is` would be the best to check this in React but I also understand that it is cheaper than quitting by checking the values ​​in depth.

> I'm not 100% sure what you mean by "missing the reference" or "lost the reference", but you just mean the object identity has changed, right?

 yeah, sorry for not being clear.